### ISSUE TITLE
Grad of Jit (and/or Jit of Grad) with custom pytrees with side effects

### ISSUE THREAD

**Issue by lfaucheux** [google/jax#18910]
### Description

In my quest for escaping tracers, I encountered something a bit 'borderline' and would like your feedback regarding wether the following behavior is intended or not.

Say you have `CPT`, a custom pytree defined as

```python
import jax

class CPT:

    __slots__ = (
        'a',
        'b',
        'c'
    )
    
    def __init__(self, *, a, b):
        self.a = a
        self.b = b

    def set_c(self, e, t=None):
        self.c = self.a * self.b * e
        return t

    def get_d(self, t=None):
        return self.c*2 + 1

    def flatten(self):
        return ((self.a, self.b), (['a', 'b'],))
        
    @classmethod
    def unflatten(cls, aux, vals):
        return cls(
            **dict(zip(aux[0], vals))
        )

    def __repr__(self):
        return (
            f'{self.__class__}'
            f'[a={self.a}, b={self.b}, '
            f'c={getattr(self, "c", None)}]'
        )

jax.tree_util.register_pytree_node(
    CPT, CPT.flatten, CPT.unflatten
)

```
... and `f`, its jitted version, its gradient, its jitted gradient, etc.. that look like

```python
def f(x, e, cpt):
    """A convoluted function I could refer to if necessary."""
    e = e.at[1:2].set(x)
    t = cpt.set_c(e, jax.lax.create_token())  # Instead of None
    d = cpt.get_d(t)
    return d.sum()

jf   = jax.jit(f)
gf   = jax.grad(f, argnums=0)
jgf  = jax.jit(gf)
gjf  = jax.grad(jf, argnums=0)
jgjf = jax.jit(gjf)
```

Let's now concretize the discussion by first considering some commonalities and eventually, the questioned behavior.

```python
from jax import numpy as jnp

_1s  = jnp.ones(4)
_2s  = 2 * _1s
_3s  = 3 * _1s
_4s  = 4 * _1s
cpt  = CPT(a=_2s, b=_3s)
args = (.5, _4s, cpt)

jax.config.update("jax_log_compiles", False)
jax.config.update("jax_check_tracer_leaks", True)

print('f(.)    [***]', f(*args))
print('jf(.)   [***]', jf(*args))

try:
    print('gf(.)  ', gf(*args))
except Exception as e:
    print(f'gf(.)   [!!!] {e!s:.35}...')

try:
    print('jgf(.) ', jgf(*args))
except Exception as e:
    print(f'jgf(.)  [!!!] {e!s:.35}...')

print('gjf(.)  [???]', gjf(*args), 'false negative ???')
print('jgjf(.) [???]', jgjf(*args), 'false negative ???')

```
which displays
```
f(.)    [***] 154.0
jf(.)   [***] 154.0
gf(.)   [!!!] Leaked trace MainTrace(2,JVPTrace)....
jgf(.)  [!!!] Leaked trace MainTrace(3,JVPTrace)....
gjf(.)  [???] 12.0 false negative ???
jgjf(.) [???] 12.0 false negative ???
```

## Questions

- Are the two last calls false negative ?

- Is `f` pure ? I would tend to say that, yes it is pure, since the computation body of `f`, however convoluted it may be, is (pleonastically) encapsulated. Still, as once explained in a message from jakevdp that I can no longer find : *"""registered pytrees with impure methods cannot work correctly with (grad or else) transformations"""*. 

- While tracing back the code paths given by the error messages (truncated twice above) and by `import ctypes; o = ctypes.cast(<id of interest>, ctypes.py_object); v = o.value` to further identify and manipulate the indicated objects, I never find myself in an execution frame of my own making. Instead, these are frames specific to the transformations themselves... As if `jax_check_tracer_leaks=True` itself generates the leaks it monitors. But it is likely that I am venturing into areas that I do not know well enough.

### What jax/jaxlib version are you using?

jax>=0.4.13 jaxlib>=0.4.13

### Which accelerator(s) are you using?

CPU | GPU

### Additional system info?

Python 3.10, Under Linux | Windows

**Comment by jakevdp** [comment]
> Is `f` pure?

No, `f` is most definitely not pure. Passing `cpt` to `f` will modify the object in-place, and that kind of side-effect is a hallmark of an impure function.

As for the rest of your questions: as soon as you pass an impure function to a JAX transformation, you have broken the contract of JAX transformations, and the result behavior will not be well-defined.

As for when the leak checker triggers vor not, this will depend on the implementation details of each (nested) transformation, and whether it results in a leaked tracer persisting due to being tored at `cpt.c`.

**Comment by lfaucheux** [comment]
> In each function that triggers a tracer leak error, if you print cpt.c after executing it, you will find that the attribute stores a tracer.

Not in the second tracer leak case (`c3` below). And in my (hardly reproducible) production case, not in the first case either (for some reasons I am still trying to get).

```python

cpt  = lambda: CPT(a=_2s, b=_3s)

print('f(.)    [***](c0)', f(*(.5, _4s, c0 := cpt())))
print('jf(.)   [***](c1)', jf(*(.5, _4s, c1 := cpt())))

try:
    print('gf(.)  ', gf(*(.5, _4s, c2 := cpt())))
except Exception as e:
    print(f'gf(.)   [!!!](c2) {e!s:.35}...')

try:
    print('jgf(.) ', jgf(*(.5, _4s, c3 := cpt())))
except Exception as e:
    print(f'jgf(.)  [!!!](c3) {e!s:.35}...')

print('gjf(.)  [???](c4)', gjf(*(.5, _4s, c4 := cpt())), 'false negative ???')
print('jgjf(.) [???](c5)', jgjf(*(.5, _4s, c5 := cpt())), 'false negative ???')

# f(.)    [***](c0) 154.0
# jf(.)   [***](c1) 154.0
# gf(.)   [!!!](c2) Leaked trace MainTrace(2,JVPTrace)....
# jgf(.)  [!!!](c3) Leaked trace MainTrace(3,JVPTrace)....
# gjf(.)  [???](c4) 12.0 false negative ???
# jgjf(.) [???](c5) 12.0 false negative ???
```

**Comment by jakevdp** [comment]
Ah, you're right. But in any case, if you remove the impure aspect of the core function, the tracer leaks should go away in all cases.

My main suggestion to you would be: don't spend energy trying to figure out which cases work and don't work with impure functions. That's a recipe for frustration and erroneous results. Rather, work to ensure that the functions you're passing to JAX transformations are pure.

**Comment by lfaucheux** [comment]
Thanks for the friendly suggestion but I'll have to spend energy on that. Be that as it may, what's the point of e.g. [sequencing-effects#motivation](https://jax.readthedocs.io/en/latest/jep/10657-sequencing-effects.html#motivation) if impure (jitted) functions are categorically 'out of question' (?)

As for  `c4`'s and `c5`'s (silently) leaked traces, as I understand it, you appear to 51%-rule out the possibility of false negatives.

**Comment by jakevdp** [comment]
Note that https://jax.readthedocs.io/en/latest/jep/10657-sequencing-effects.html is a design document proposing a system to track a limited set of well-defined runtime side-effects through jax transformations in order to support things like `jax.debug.print`. It's not really relevant to the type of trace-time side-effect that appears in your code.

> Regarding `f`'s unachieved purity, given that the (inplace) modification is confined entirely within its local scope

I don't agree with this statement. The side-effect of `f` is to modify object passed to it, in a way that affects behavior outside the function's scope. Consider this for example:
```python
cpt  = CPT(a=_2s, b=_3s)
print(hasattr(cpt, "c") # False
result = f(.5, _4s, cpt)
print(hasattr(cpt, "c") # True
```
This in-place mutation of the function input is what leads to the tracer leak under some transformations, and is what makes your code incompatible with JAX transformations in general.

**Comment by lfaucheux** [comment]
> This in-place mutation of the function input is what leads to the tracer leak [...]

True. And I had this in mind when writing this. What about providing `f` with (deep)copies of `CPT` instances ? Or one step further : instantiating them inside `f` ? Would `f`'s purity be achieved, at least in the sense that JAX defines and operationalizes it ?

**Comment by jakevdp** [comment]
Yes, that version is fine, because the mutated object only exists within the scope of the function. So the function is pure in the sense that JAX needs: i.e. the function only affects the Python namespace via its return values.

**Comment by lfaucheux** [comment]
Do you think I can expect a significant improvement in memory consumption and (XLA) compilation time simply by instantiating my `CPT`s inside `f` ? Especially given that `f` is 15000 lines long and contains dozens of `CPT`s. Does the size of `f` overshadow all other considerations?

**Comment by jakevdp** [comment]
For a `CPT` class like the one above, I wouldn't expect instantiating CPT inside vs. outside `f` to make much difference in compilation time, runtime, or memory use (but the baseline in this case is code that's not compatible with JAX, so it may not be a fair comparison!)

**Comment by jakevdp** [comment]
It seems like this has been answered – closing!
